src/vm/marshalnative.cpp

```diff
@@ -559,10 +558,6 @@ FCIMPL2(LPVOID, MarshalNative::GCHandleInternalAlloc, Object *obj, int type)
 
     HELPER_METHOD_FRAME_BEGIN_RET_NOPOLL();
 
-    // If it is a pinned handle, check the object type.
-    if (type == HNDTYPE_PINNED)
-        ValidatePinnedObject(objRef);
-
```

Comment:
@jkotas, I didn't remove the HELPER_METHOD_FRAME* stuff from here because it looks like at least CreateTypedHandle can throw, and thus I assume the helper frame support is needed.  However, these changes seem to result in maybe a ~10% regression in GCHandle.Alloc(…, Pinned); I'm guessing that has to do with now making two FCalls where we previously made one, and this helper frame still needing to be here.  Suggestions?  Or did I misunderstand the recommendation?

Reply:
I have refactored the FCalls to have fast path to help with this. `GCHandle.Alloc(..., GCHandleType.Pinned)/Free` in a loop is ~30% faster compared to master for me now.